web/modules/custom/dpl_related_content/dpl_related_content.module

```diff
@@ -0,0 +1,292 @@
+ * inheritance overriding in favor of a simpler codebase.
+ *
+ * @param string|null $current_uuid
+ *   A possible entity UUID, that will not get included in results.
+ * @param array<int> $tags
+ *   Tag term IDs, to look for.
+ * @param array<int> $categories
+ *   Category term IDs, to look for.
+ * @param array<int> $branches
+ *   Branch term IDs, to look for.
+ *
+ * @return array<int|string>
+ *   Matching eventinstance IDs
+ *
+ * @throws \Exception
+ */
+function _dpl_related_content_get_eventinstance_ids(?string $current_uuid, array $tags = [], array $categories = [], array $branches = []): array {
+  $date = new DrupalDateTime('today');
+  $date->setTimezone(new \DateTimezone(DateTimeItemInterface::STORAGE_TIMEZONE));
+  $formatted_date = $date->format(DateTimeItemInterface::DATETIME_STORAGE_FORMAT);
+
+  $connection = Database::getConnection();
+
+  // Prepare a subquery for eventseries IDs based on terms.
+  $subquery = $connection->select('eventseries', 'es');
+  $subquery->fields('es', ['id']);
+
+  // Join the term field tables.
+  // @todo this only looks at the values on the eventseries.
+  // Ideally, we'd want to also look if values exist on the instance level.
+  $subquery->leftJoin('eventseries__field_tags', 'es_tags', 'es.id = es_tags.entity_id');
+  $subquery->leftJoin('eventseries__field_categories', 'es_cats', 'es.id = es_cats.entity_id');
+  $subquery->leftJoin('eventseries__field_branch', 'es_bra', 'es.id = es_bra.entity_id');
+
+  if (!empty($tags) || !empty($categories) || !empty($branches)) {
+    $or_group = $subquery->orConditionGroup();
+
+    // To avoid errors, related to the OR GROUP only containing one condition,
+    // we'll add a fake condition to fill out a possible empty space.
+    $or_group->condition('title', 'ALWAYS_FALSE');
+
+    if (!empty($tags)) {
+      $or_group->condition('es_tags.field_tags_target_id', $tags, 'IN');
+    }
+
+    if (!empty($categories)) {
+      $or_group->condition('es_cats.field_categories_target_id', $categories, 'IN');
+    }
+
+    if (!empty($branches)) {
+      $or_group->condition('es_bra.field_branch_target_id', $branches, 'IN');
+    }
+
+    $subquery->condition($or_group);
+  }
+
+  $subquery->distinct(TRUE);
+
+  // Main query to select eventinstance ids, joining with
+  // eventinstance_field_data for condition fields.
+  $query = $connection->select('eventinstance_field_data', 'eid');
+  $query->join('eventinstance', 'ei', 'ei.id = eid.id');
+  $query->addField('eid', 'id', 'eventinstance_id');
+
+  if (!empty($tags) || !empty($categories) || !empty($branches)) {
+    // Use the subquery to filter by eventseries_id.
+    $query->condition('eid.eventseries_id', $subquery, 'IN');
+  }
+
+  $query->condition('ei.uuid', $current_uuid, '<>');
+
+  // We only want events in the future (e.g. - active events).
+  $query->condition('eid.date__value', $formatted_date, '>=');
+  $query->orderBy('eid.date__value', 'ASC');
+  $query->range(0, 16);
```

Comment:
Please add a comment explaining why this range is necessary.

In general I would have expected us to include all the matching ids. If we wanted to do any limiting we could do it later on in the prcoess.

Reply:
There's no point in pulling more than what will be the maximum allowed in the slider, but, i'll add documentation and a variable to make it clearer.